README.md
nfldb is a relational database bundled with a Python module to quickly and 
conveniently query and update the database with data from active games.
Data is imported from
[nflgame](https://github.com/BurntSushi/nflgame), which in turn gets its data
from a JSON feed on NFL.com's live GameCenter pages. This data includes, but is 
not limited to, game schedules, scores, rosters and play-by-play data for every 
preseason, regular season and postseason game dating back to 2009.

It can also
[be used with nflvid to search for and watch video of 
plays](http://pdoc.burntsushi.net/nflvid.vlc). Please see the
[nfldb wiki](https://github.com/BurntSushi/nfldb/wiki) for more details on how 
to get that working.

Here is a small teaser that shows how to use nfldb to find the top five passers 
in the 2012 regular season:

```python
import nfldb

db = nfldb.connect()
q = nfldb.Query(db)

q.game(season_year=2012, season_type='Regular')
for pp in q.sort('passing_yds').limit(5).as_aggregate():
    print pp.player, pp.passing_yds
```

And the output is:

```bash
[andrew@Liger ~] python2 top-five.py 
Drew Brees (NO, QB) 5177
Matthew Stafford (DET, QB) 4965
Tony Romo (DAL, QB) 4903
Tom Brady (NE, QB) 4799
Matt Ryan (ATL, QB) 4719
```


### Documentation and getting help

nfldb has
[comprehensive API documentation](http://pdoc.burntsushi.net/nfldb). 
Tutorials, more examples and a description of the data model can be found
on the [nfldb wiki](https://github.com/BurntSushi/nfldb/wiki).

If you need any help or have found a bug, please
[open a new issue on nfldb's issue 
tracker](https://github.com/BurntSushi/nfldb/issues/new)
or join us at our IRC channel `#nflgame` on FreeNode.


### Installation and dependencies

nfldb depends on the following Python packages available in
[PyPI](https://pypi.python.org/pypi):
[nflgame](https://pypi.python.org/pypi/nflgame),
[psycopg2](https://pypi.python.org/pypi/psycopg2),
[pytz](https://pypi.python.org/pypi/pytz) and
[enum34](https://pypi.python.org/pypi/enum34).
nfldb also needs PostgreSQL installed with an available empty database.

I've only tested nfldb with Python 2.7 on a Linux system. In theory, nfldb 
should be able to work on Windows and Mac systems as long as you can get 
PostgreSQL running. It is not Python 3 compatible (yet, mostly because of
the `nflgame` dependency).

Please see the
[installation guide](https://github.com/BurntSushi/nfldb/wiki/Installation)
on the [nfldb wiki](https://github.com/BurntSushi/nfldb/wiki)
for instructions on how to setup nfldb.


### Entity-relationship diagram

Here's a condensed version that excludes play and player statistics:

[![Shortened ER diagram for nfldb](http://burntsushi.net/stuff/nfldb/nfldb-condensed.png)](http://burntsushi.net/stuff/nfldb/nfldb-condensed.pdf)

There is also a [full PDF version](http://burntsushi.net/stuff/nfldb/nfldb.pdf)
that includes every column in the database.

The nfldb wiki has a [description of the data 
model](https://github.com/BurntSushi/nfldb/wiki/The-data-model).

The most recent version of the nfldb PostgreSQL database is available here:
[http://burntsushi.net/stuff/nfldb/nfldb.sql.zip](http://burntsushi.net/stuff/nfldb/nfldb.sql.zip).



CHANGELOG
0.2.17
======
- Fixed nfldb.team.standard_team to recognize LA Rams.


0.2.16
======
- Updates for the 2016 season.
- Includes database migration to add new LA team.


0.2.15
======
- Revert PR #142.


0.2.14
======
- Fix #134 by bumping nflgame dependency.


0.2.11
======
- Add custom override for start time for 2015100400


0.2.10
======
- Fix #80.


0.2.9
=====
- Update nflgame dependency to get playoff games.


0.2.8
=====
- Fix time handling for games starting at 12pm/12:30pm.
- Depend on nflgame >= 1.2.9.


0.2.7
=====
- Depend on nflgame >= 1.2.8 (see nflgame#95).


0.2.6
=====
- Fixed a bug that was preventing `nfldb.aggregate` from working.


0.2.5
=====
- Added a new "simulation" mode to `nfldb-update`. Try it with
  `nfldb-update --simulate 2014081501`.
- Added a new `game_date` field to plays that corresponds to the
  first 8 characters of the play's game gsis_id. This ends up
  translating to the date that the game played, which just so happens
  to be how play identifiers are handed out on NFL.com. This makes
  sorting plays with respect to *when* they happened simple.
- Refactored some of nfldb-update into its own module so it can
  be invoked easier from other programs.
- Better error messages when sorting with invalid field.


0.2.4
=====
Minor refactoring and performance features.

The refactoring is mostly confined to the SQL 'ORDER BY' generator. I
had intended to make it more powerful, but that turned out to be more
work than I expected. Instead, I'll settle for clearer code.

The performance features aren't likely to be useful unless you're doing
heavy analysis or are using nfldb to power a web site. Briefly, several
`fill` functions have been added to nfldb's entity types that will, for
example, fill in a list of plays with drive and game data using as few
queries as possible. Currently, this will fall over with >5000 items
because of PostgreSQL limits, but future work will fix this using some
sort of batching mechanism.


0.2.3
=====
- Fixes #41 (Python 2.6 compatibility).


0.2.2
=====
- Fixes #39 and #42.
- Fixes another bug where LIMIT clauses weren't being inserted by
  `nfldb.Query` if there were no sorting criteria.


0.2.1
=====
- Fixes #34.


0.2.0
=====
Materialized view for `play`. `Query` demolition.

ATTN: This introduces a breaking change. The `team` field can no longer
be used in the `play` method. Instead, you should use the new
`play_player` method to select individual player statistics belonging to
a specific team.

Otherwise, there are very few public facing changes, but the entire
guts of `nfldb.Query` have been ripped out and replaced with more
robust SQL generation code. Moreover, several idiosyncracies have been
fixed and some unit tests have finally been added.

1. Previously, the `Query` class was doing some very clever things to do
   parts of a JOIN in Python code. The general flow was that filtering
   was applied to find primary keys---never using any JOINs---and once
   all criteria had been applied, those ids were used in a simple SELECT
   to fetch the actual rows.

   Now all of that cruft has been removed and replaced with intelligent
   SQL generation that constructs one query with all the proper JOINs.
   For whatever reason, I thought this was slower when experimenting
   with it when I first started nfldb. Perhaps my indexes weren't
   configured properly then. In any case, I can't really see much
   performance difference.

2. The SQL generation code is very smart. Although it is not part of
   nfldb's public API, I imagine it would be very useful if you had some
   special needs. See the unexported but documented `nfldb.sql` module.

3. Many idiosyncracies resulting from doing a join in Python are now
   completely gone. For example, if you tried to apply a `sort` with a
   `limit` with complex search criteria, you were bound to get wrong
   answers. For example, if you tried sorting by both a column on the
   `week` table (like `down`) and a column on `play_player` (like
   `passing_tds`) and applied a limit to it, the results would be
   completely wonky because the pure Python join can't cope with it
   performantly. A regular SQL join? Piece of cake.

4. I have added a materialized view `agg_play`. This is a fancy word for
   "a table that automatically updates itself." In essence, whenever a
   new row is added to `play_player`, aggregate statistics for that play
   are re-computed. This makes adding data slower (which doesn't happen
   very frequently), but it makes querying data much faster and easier.
   For example, plays can be queried for `passing_yds` without ever
   joining with `play_player`. (Which is wonky because of the
   one-to-many relationship.)
   To reflect this clearer separation of concerns, the `Query.play`
   method will no longer add criteria that hits the `play_player` table.
   Instead, if you really want the `play_player` table, then you can use
   the new `play_player` method. The only field that was accepted in the
   `play` that is no longer allowed is the `team` and `player_id`
   fields. This is because there is no sensible way to aggregate these
   values into a single play.

   To the best of my knowledge, that is the only possible breaking
   change here.


0.1.6
=====
  - Add better error message when config file can't be found.

0.1.5
=====
  - Support hall-of-fame games by allowing week to be `0`.

0.1.4
=====
  - Some light code refactoring and doco updates.

0.1.3
=====
  - Fix a bug where the start time of a game has the wrong
    year if it started the year after the season started.
    (e.g., postseason games.)

0.1.2
=====
  - nflgame now correctly stores the season of postseason
    games, so nfldb no longer needs to fix it.

0.1.1
=====
  - Added a fix so that nfldb can work with psycopg2 2.4.x.

0.1.0
=====
  - Supports nflgame's new automatic schedule updating.
    (nfldb now automatically updates schedules.)

0.0.16
======

nfldb/__init__.py
"""
Module nfldb provides command line tools and a library for maintaining
and querying a relational database of play-by-play NFL data. The data
is imported from [nflgame](https://github.com/BurntSushi/nflgame),
which in turn gets its data from a JSON feed on NFL.com's live
GameCenter pages. This data includes, but is not limited to, game
schedules, scores, rosters and play-by-play data for every preseason,
regular season and postseason game dating back to 2009.

Here is a small teaser that shows how to use nfldb to find the top five
passers in the 2012 regular season:

    #!python
    import nfldb

    db = nfldb.connect()
    q = nfldb.Query(db)

    q.game(season_year=2012, season_type='Regular')
    for pp in q.sort('passing_yds').limit(5).as_aggregate():
        print pp.player, pp.passing_yds

And the output is:

    [andrew@Liger ~] python2 top-five.py
    Drew Brees (NO, QB) 5177
    Matthew Stafford (DET, QB) 4965
    Tony Romo (DAL, QB) 4903
    Tom Brady (NE, QB) 4799
    Matt Ryan (ATL, QB) 4719

In theory, both `nfldb` and `nflgame` provide access to the same data.
The difference is in the execution. In order to search data in nflgame,
a large JSON file needs to be read from disk and loaded into Python
data structures for each game. Conversely, nfldb's data is stored in
a relational database, which can be searched and retrieved faster
than nflgame by a few orders of magnitude. Moreover, the relational
organization of data in nfldb allows for a convenient
[query interface](http://goo.gl/Sd6MN2) to search NFL play data.

The database can be updated with real time data from active games by
running the `nfldb-update` script included with this module as often
as you're comfortable pinging NFL.com. (N.B. The JSON data itself only
updates every 15 seconds, so running `nfldb-update` faster than that
would be wasteful.) Roster updates are done automatically at a minimum
interval of 12 hours.

nfldb has [comprehensive API documentation](http://pdoc.burntsushi.net/nfldb)
and a [wiki with examples](https://github.com/BurntSushi/nfldb/wiki).

nfldb can be used in conjunction with
[nflvid](https://pypi.python.org/pypi/nflvid)
to
[search and watch NFL game footage](http://goo.gl/Mckaf0).

If you need help, please join us at our IRC channel `#nflgame` on
FreeNode.
"""
from __future__ import absolute_import, division, print_function

from nfldb.db import __pdoc__ as __db_pdoc__
from nfldb.db import api_version, connect, now, set_timezone, schema_version
from nfldb.db import Tx
from nfldb.query import __pdoc__ as __query_pdoc__
from nfldb.query import aggregate, current, guess_position, player_search
from nfldb.query import Query, QueryOR
from nfldb.team import standard_team
from nfldb.types import __pdoc__ as __types_pdoc__
from nfldb.types import stat_categories
from nfldb.types import Category, Clock, Enums, Drive, FieldPosition, Game
from nfldb.types import Play, Player, PlayPlayer, PossessionTime, Team
from nfldb.version import __pdoc__ as __version_pdoc__
from nfldb.version import __version__

__pdoc__ = __db_pdoc__
__pdoc__ = dict(__pdoc__, **__query_pdoc__)
__pdoc__ = dict(__pdoc__, **__types_pdoc__)
__pdoc__ = dict(__pdoc__, **__version_pdoc__)


# Export selected identifiers from sub-modules.
__all__ = [
    # nfldb.db
    'api_version', 'connect', 'now', 'set_timezone', 'schema_version',
    'Tx',

    # nfldb.query
    'aggregate', 'current', 'guess_position', 'player_search',
    'Query', 'QueryOR',

    # nfldb.team
    'standard_team',

    # nfldb.types
    'select_columns', 'stat_categories',
    'Category', 'Clock', 'Enums', 'Drive', 'FieldPosition', 'Game',
    'Play', 'Player', 'PlayPlayer', 'PossessionTime', 'Team',

    # nfldb.version
    '__version__',
]
